### ISSUE TITLE
머딥 - 17.6d (2층 신경망 직접 구현)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#53]
# 신경망 구현
**복습할 점**
1. Class 안에 self.params  dictionary 테크닉 숙지하기
2. predict 안에서 파라미터 초기값 unpacking 하기
3. numerical_gradient에서 key값 params과 맞추기
4. tqdm_notebook 단계 정보확인
5. mini batch index? mask
6. **forward propagation 미분이 느린 이유**? 신경망은 param.이 많기 때문

```
from tensorflow.keras import datasets 
import numpy as np

mnist = datasets.mnist
(X_train, t_train), (X_test, t_test) = mnist.load_data()

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def softmax(x):
    if x.ndim == 2:
        x = x.T
        x = x - np.max(x, axis=0)
        y = np.exp(x) / np.sum(np.exp(x), axis=0)
        return y.T 
    x = x - np.max(x) # 오버플로 대책
    return np.exp(x) / np.sum(np.exp(x))

def cross_entropy_error(y, t):
    if y.ndim == 1:
        t = t.reshape(1, t.size)
        y = y.reshape(1, y.size)
    # 훈련 데이터가 원-핫 벡터라면 정답 레이블의 인덱스로 반환
    if t.size == y.size:
        t = t.argmax(axis=1)
    batch_size = y.shape[0]
    return -np.sum(np.log(y[np.arange(batch_size), t] + 1e-7)) / batch_size

def numerical_gradient(f, x):
    h = 1e-4 # 0.0001
    grad = np.zeros_like(x)
    it = np.nditer(x, flags=['multi_index'], op_flags=['readwrite'])
    while not it.finished:
        idx = it.multi_index
        tmp_val = x[idx]
        x[idx] = float(tmp_val) + h
        fxh1 = f(x) # f(x+h)
        x[idx] = tmp_val - h 
        fxh2 = f(x) # f(x-h)
        grad[idx] = (fxh1 - fxh2) / (2*h)
        x[idx] = tmp_val # 값 복원
        it.iternext()   
    return grad
```

```sage
class TwoLayerNet:    
    def __init__(self, input_size=28*28, hidden_size=100, output_size=10, weight_init_std=0.01):
        self.params = {}
        self.params['W1'] = np.random.randn(input_size, hidden_size) * weight_init_std
        self.params['b1'] = np.zeros(hidden_size)
        self.params['W2'] = np.random.randn(hidden_size, output_size) * weight_init_std
        self.params['b2'] = np.zeros(output_size)
                
    def predict(self, x):
        # 마지막 출력층에서 softmax 사용할 것
        W1, b1 = self.params["W1"], self.params["b1"]
        W2, b2 = self.params["W2"], self.params["b2"]

        # 1층 은닉층 계산
        z1 = x @ W1 + b1
        a1 = sigmoid(z1)

        # 2층 출력층 계산
        z2 = a1 @ W2 + b2
        y = softmax(z2)
        return y
        
    def loss(self, x, t):
        y = self.predict(x)
        loss = cross_entropy_error(y, t)
        return loss
        
    def numerical_gradient_params(self, x, t):
        grad = {}
        g = lambda p : self.loss(x, t)
        
        grad['W1'] = numerical_gradient(g, self.params['W1'])
        grad['b1'] = numerical_gradient(g, self.params['b1'])
        grad['W2'] = numerical_gradient(g, self.params['W2'])
        grad['b2'] = numerical_gradient(g, self.params['b2'])
        return grad
```

**OneHot Encoding - Tensorflow**
```sage
import tensorflow as tf

t_train_oh = tf.one_hot(t_train, 10).numpy()
t_test_oh = tf.one_hot(t_test, 10).numpy()

X_train = X_train.reshape(X_train.shape[0], -1)
X_test = X_test.reshape(X_test.shape[0], -1)
```

## 이미지 데이터 정규화
```
X_train = X_train/255.  # image data scaling
X_test = X_test/255.

iter_nums = 10000
train_size = X_train.shape[0]        
batch_size = 100
learning_rate = 0.1

net = TwoLayerNet()
```

### tqdm
```sage
from tqdm import tqdm_notebook

for i in tqdm_notebook(range(iter_nums)):
    # 미니배치 인덱스 설정(mask)
    batch_mask = np.random.choice(train_size, batch_size)
    
    # 미니배치 만들기
    X_batch = X_train[batch_mask]
    t_batch = t_train[batch_mask]
    
    # 파라미터 기울기 구하기(=학습)
    grad = net.numerical_gradient_params(X_batch, t_batch)
    
    # 경사하강법 (param과 grad가 키를 공유하면 편함)
    keys = ['W1', 'b1', 'W2', 'b2']
    
    # 경사하강법이란 파라미터를 업데이트 하는 것
    for key in keys:
        net.params[key] -= learning_rate * grad[key]
    
    loss = net.loss(X_batch, t_batch)
    print('Step {} -> Loss : {}'.format(i+1, loss))
```
